[WIP] Introduce BigInt Public C API in JavaScriptCore

hyjorc1 opened the pull request:
#### c97a08d9e56ffc23e2f770be827909d2caa64a22
<pre>
[WIP] Introduce BigInt Public C API in JavaScriptCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=250511">https://bugs.webkit.org/show_bug.cgi?id=250511</a>
<a href="https://rdar.apple.com/104194532">rdar://104194532</a>

Reviewed by Keith Miller.

JavaScriptCore framework provides the ability to evaluate JavaScript programs
within applications developed using Swift and Object-C. While JavaScriptCore
APIs support all JavaScript primitive types except for BigInt type. This patch
introduces BigInt public C APIs for type verification, type access, value
creation, value conversion, and value comparison.

* Source/JavaScriptCore/API/APICast.h:
(toRef):
(toBigIntRef):
* Source/JavaScriptCore/API/JSBase.h:
* Source/JavaScriptCore/API/JSBigIntRef.cpp: Added.
(JSBigIntCreateWithNumber):
(JSBigIntCreateWithString):
(JSBigIntToString):
* Source/JavaScriptCore/API/JSBigIntRef.h: Added.
* Source/JavaScriptCore/API/JSValue.h:
* Source/JavaScriptCore/API/JSValue.mm:
(-[JSValue isBigInt]):
* Source/JavaScriptCore/API/JSValueRef.cpp:
(JSValueGetType):
(JSValueIsBigInt):
(JSValueMakeBigInt):
(JSValueToBigInt):
* Source/JavaScriptCore/API/JSValueRef.h:
* Source/JavaScriptCore/API/JavaScript.h:
* Source/JavaScriptCore/API/tests/testapi.cpp:
(TestAPI::checkIsBigIntType):
(TestAPI::testBigInt):
(testCAPIViaCpp):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::JSBigInt::stringToBigInt):
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::asBigInt const):

Canonical link: <a href="https://commits.webkit.org/278275@main">https://commits.webkit.org/278275@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39f1587904703f7cf16cdce2a3f258ba1c879dc0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50071 "34 style errors") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29362 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2357 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53321 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/753 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35534 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/317 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52168 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/27005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43091 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21961 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/333 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8450 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/43396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/383 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54908 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49569 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25165 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26424 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/2357 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10978 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Checked out pull request; Reviewed by Keith Miller; Compiled WebKit (warnings); Running layout-tests; Canonicalized commit; Pushed commit to WebKit repository") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/27285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26156 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11718 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/API/JSBigIntRef.h:
```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2023 Apple Inc.  All rights reserved.
```
2023 -> 2024

hyjorc1 commented on Source/JavaScriptCore/API/JSBigIntRef.cpp:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
2023 -> 2024

kmiller68 commented on Source/JavaScriptCore/Sources.txt:
```diff
@@ -25,6 +25,7 @@ API/APIIntegrity.cpp
 API/JSAPIGlobalObject.cpp
 API/JSAPIValueWrapper.cpp
 API/JSBase.cpp
+API/JSBigIntRef.cpp @no-unify
```
Why is this `@no-unify`?

hyjorc1 replied:
There is a ambiguity problem due to `JSType` is defined in both `JSType.h` and `JSValueRef.h`.

kmiller68 replied:
I think the real problem here is that we do `using namespace JSC` in our API .cpp files. This is the problem with not having scoped namespace expansion...

hyjorc1 replied:
The problem is not caused in `JSBigIntRef.cpp`. With the new file JSBigIntRef.cpp added, a unified source file can have this weird link path:
```
In file included from /Volumes/WebKit/macOS/OpenSource/WebKitBuild/Debug/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource4.cpp:2:
In file included from ./API/JSTypedArray.cpp:30:
In file included from ./API/APICast.h:31:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSCJSValueInlines.h:35:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSCellInlines.h:32:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/interpreter/CallFrameInlines.h:29:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSCalleeInlines.h:29:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/StructureInlines.h:33:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSObjectInlines.h:32:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSGenericTypedArrayViewInlines.h:30:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/GenericTypedArrayViewInlines.h:29:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:31:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/API/JSClassRef.h:32:
/Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/API/JSObjectRef.h:280:73: error: reference to 'JSType' is ambiguous
(*JSObjectConvertToTypeCallback) (JSContextRef ctx, JSObjectRef object, JSType type, JSValueRef* exception);
                                                                        ^
In file included from /Volumes/WebKit/macOS/OpenSource/WebKitBuild/Debug/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource4.cpp:1:
In file included from ./API/JSStringRef.cpp:27:
In file included from ./API/JSStringRef.h:29:
/Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/API/JSValueRef.h:56:3: note: candidate found by name lookup is 'JSType'
} JSType;
  ^
In file included from /Volumes/WebKit/macOS/OpenSource/WebKitBuild/Debug/DerivedSources/JavaScriptCore/unified-sources/UnifiedSource4.cpp:2:
In file included from ./API/JSTypedArray.cpp:30:
In file included from ./API/APICast.h:29:
In file included from ./API/JSAPIValueWrapper.h:25:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/interpreter/CallFrame.h:30:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/VM.h:40:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/IndexingType.h:29:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/bytecode/SpeculatedType.h:33:
In file included from /Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/TypedArrayType.h:28:
/Volumes/WebKit/macOS/OpenSource/Source/JavaScriptCore/runtime/JSType.h:138:6: note: candidate found by name lookup is 'JSC::JSType'
enum JSType : uint8_t {
     ^
1 error generated.
```

kmiller68 commented on Source/JavaScriptCore/API/JSBase.h:
```diff
@@ -68,6 +68,9 @@ typedef const struct OpaqueJSValue* JSValueRef;
 typedef struct OpaqueJSValue* JSObjectRef;
 
+/*! @typedef JSBigIntRef A JavaScript bigint. A JSBigInt is a JSValue. */
+typedef const struct OpaqueJSValue* JSBigIntRef;
```
I guess I don't really get the point of having a JSBigIntRef type. You can't do anything with it that you can't do with the JSValue. We don't really have special types for other primitive types with the exception of JSStringRef, which gives you back something other than a JSValue anyway. Can you explain your thought process here?

hyjorc1 replied:
I think there are several reasons:
1. Unlike other primitives (except for JSString) which can be directly converted to C types.
2. With JSBigIntRef, no need to check whether it's bigint or not for its related APIs (JSBigIntToInt64, JSBigIntToUInt64, .etc).
3. JSBigIntRef follows modular API design which breaks down JSValue into smaller and self-contained components which can be developed, deployed, and maintained independently. Bigint arithmetic operations, which requires APIs due to (1), is not introduced yet but might be added on demands. If those operations are needed, then JSBigIntRef is beneficial since (2).

kmiller68 replied:
I don't think I agree with your arguments here. As this API stands right now, JSBigIntRef adds a lot of surface area but I don't see how it makes life more convenient for users of the API. If we decide to add arithmetic operations in the future we can always add JSBigIntRef at that point. I'm also not convinced that eliminating the checks is a significant win. Users already have to call a function to do the addition, which likely involves an object allocation. Lastly, there are primitives that don't have a C type other than JSString, namely JSSymbol, which don't have a dedicated type. 

IMO, if we want to have a more modular API we should probably do it in the Obj-C/Swift API first since that API is more widely used and it actually has proper subclassing. Also, we should do it more comprehensively with many subclasses for many different types.

kmiller68 commented on Source/JavaScriptCore/API/JSBigIntRef.h:
```diff
@@ -0,0 +1,114 @@
+    @param value      The JSBigInt to Truncate.
+    @result           A 64-bit signed integer representing the JSBigInt value.
+*/
+JS_EXPORT int64_t JSBigIntToBigInt64(JSContextRef ctx, JSBigIntRef value) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
I think this should be `JSBigIntToInt64` for symmetry with the inverse constructor.

kmiller68 commented on Source/JavaScriptCore/API/JSBigIntRef.h:
```diff
@@ -0,0 +1,114 @@
+    @param value      The JSBigInt to Truncate.
+    @result           A 64-bit unsigned integer representing the JSBigInt value.
+*/
+JS_EXPORT uint64_t JSBigIntToBigUInt64(JSContextRef ctx, JSBigIntRef value) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
I think this should be `JSBigIntToUInt64` for symmetry with the inverse constructor.

kmiller68 commented on Source/JavaScriptCore/API/JSBigIntRef.h:
```diff
@@ -0,0 +1,114 @@
+    @param ctx        The execution context to use.
+    @param number     The number to copy into the new JSBigInt.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A JSBigInt representing the value of the number, or NULL if an exception is thrown.
```
We should probably add `@discussion If number is not an integer this function will return an exception.`

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -171,6 +171,33 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a unique JavaScript value with type bigint.
+*/
++ (JSValue *)valueWithNewBigIntFromUInt64:(uint64_t)uint64 inContext:(JSContext *)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
How come we don't have a number constructor like we do for the C API?

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +302,35 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+ @param exception     A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+ @result              A JSValue of the bigint type, representing the value of bigint. Or NULL if an exception is thrown.
+ */
+JS_EXPORT JSValueRef JSValueMakeBigUInt64(JSContextRef ctx, JSBigIntRef bigint, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Ditto.

hyjorc1 replied:
It's requested in https://bugs.webkit.org/show_bug.cgi?id=250719.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +302,35 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+ @param exception     A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+ @result              A JSValue of the bigint type, representing the value of bigint. Or NULL if an exception is thrown.
+ */
+JS_EXPORT JSValueRef JSValueMakeBigInt64(JSContextRef ctx, JSBigIntRef bigint, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Is this really a common enough operation that it warrants API? I don't see myself using this much, if ever...

hyjorc1 replied:
It's requested in https://bugs.webkit.org/show_bug.cgi?id=250719.

kmiller68 replied:
But this function doesn't make any sense. The definition of JSBigIntRef implies that it can be used as a JSValueRef. I also don't think this is the function they are asking for, nor does them asking for this mean we have to provide it.
```
/*! @typedef JSBigIntRef A JavaScript bigint. A JSBigInt is a JSValue. */
typedef const struct OpaqueJSValue* JSBigIntRef;
```

hyjorc1 replied:
I am thinking what if we want to use `JSObjectSetProperty` to set a property with a bigint value. Then, we still need `JSValueRef` form of a `JSBigIntRef`, right?
```
JS_EXPORT void JSObjectSetProperty(JSContextRef ctx, JSObjectRef object, JSStringRef propertyName, JSValueRef value, JSPropertyAttributes attributes, JSValueRef* exception);
```

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+        return nullptr;
+    }
+
+    ASSERT(jsValue.isBigInt());
```
Nit: I don't think we need this ASSERT given we do a check right above.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @function
+    @abstract         Truncates a JavaScript bigint to an unsigned 64-bit integer and returns the resulting integer.
+    @param ctx        The execution context to use.
+    @param value      The bigint JSValue to Truncate.
```
Truncate => truncate

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @function
+    @abstract         Truncates a JavaScript bigint to a singed 64-bit integer and returns the resulting integer.
+    @param ctx        The execution context to use.
+    @param value      The bigint JSValue to Truncate.
```
Truncate => truncate

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @function
+    @abstract         Convert a bigint JSValue to JSString.
+    @param ctx        The execution context to use.
+    @param value      The bigint JSValue to Truncate.
```
Truncate => convert to string.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @abstract         Convert a bigint JSValue to JSString.
+    @param ctx        The execution context to use.
+    @param value      The bigint JSValue to Truncate.
+    @param radix      The integer in the range from 2 to 36 as the base for representing the bigint JSValue value.
```
An integer in the range 2 to 36 to be used as the base for representing the bigint value.

kmiller68 replied:
Also how did we pick that range?

hyjorc1 replied:
> radix: An integer in the range 2 through 36 specifying the base to use for representing the BigInt value. Defaults to 10.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toString#radix

kmiller68 replied:
Whoops, I meant to delete that comment when I realized it was part of the JS spec.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param radix      The integer in the range from 2 to 36 as the base for representing the bigint JSValue value.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A JSString representing the bigint JSValue value, or NULL if an exception is thrown.
+    @discussion       If value is not an bigint this function will return an exception.
```
If value is not an bigint or radix is not between 2 and 36 this function will return an exception.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+
+    JSValue result = JSBigInt::makeHeapBigIntOrBigInt32(globalObject, number);
+    if (exception)
+        *exception = nullptr;
```
We normally don't touch exception unless something was thrown.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+
+    JSValue result = JSBigInt::makeHeapBigIntOrBigInt32(globalObject, integer);
+    if (exception)
+        *exception = nullptr;
```
ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+
+    JSValue result = JSBigInt::makeHeapBigIntOrBigInt32(globalObject, integer);
+    if (exception)
+        *exception = nullptr;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+
+    JSValue result = JSBigInt::parseInt(globalObject, string->string(), JSBigInt::ErrorParseMode::ThrowExceptions);
+    if (exception)
+        *exception = nullptr;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+    JSValue jsValue = toJS(globalObject, value);
+
+    if (exception)
+        *exception = nullptr;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+
+    JSValue jsValue = toJS(globalObject, value);
+    if (exception)
+        *exception = nullptr;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+    ASSERT(jsValue.isBigInt());
+    String result = jsValue.asBigInt(globalObject)->toString(globalObject, radix);
+    if (exception)
+        *exception = nullptr;
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.cpp:
```diff
@@ -368,6 +385,174 @@ JSValueRef JSValueMakeSymbol(JSContextRef ctx, JSStringRef description)
+    }
+
+    ASSERT(jsValue.isBigInt());
+    String result = jsValue.asBigInt(globalObject)->toString(globalObject, radix);
```
I think this should be 
```
    String result;
#if USE(BIGINT32)
    if (jsValue.isBigInt32())
        result = String::number(jsValue.bigInt32AsInt32());
    else
#endif
        result = jsValue.asHeapBigInt()->toString(globalObject, radix);
```
Then you don't need `asBigInt()` and we don't allocate an extra object for int32 bigints.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @function
+    @abstract         Creates a JavaScript bigint with an integer represented in string.
+    @param ctx        The execution context to use.
+    @param string     The JSString representation of an integer.
```
JSString => JSStringRef

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -171,6 +171,43 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Create a new, unique, bigint object.
+@param string The string representation of the bigint object being created.
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a unique JavaScript value with type bigint.
```
Might be worth saying ```@discussion This is equivalent to calling the `BigInt` constructor from JavaScript with a string argument.```

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +304,81 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param ctx        The execution context to use.
+    @param string     The JSString representation of an integer.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A bigint JSValue of the string, or NULL if an exception is thrown.
```
Might be worth saying ```@discussion This is equivalent to calling the `BigInt` constructor from JavaScript with a string argument.```

kmiller68 commented on Source/JavaScriptCore/API/tests/testapi.cpp:
```diff
@@ -742,6 +757,258 @@ void TestAPI::testJSObjectSetOnGlobalObjectSubclassDefinition()
+    }
+
+    {
+        const char* description = "checking JSBigIntToString with BigInt(0x2a) and toString(16)";
```
Nit: This descriptions could probably be updated.

hyjorc1 replied:
This description should be correct. But I did update other descriptions for testing `ReferenceError: not a bigint JSValue`.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -171,6 +171,44 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@result The JSValue representing a unique JavaScript value with type bigint.
+@discussion This is equivalent to calling the `BigInt` constructor from JavaScript with a string argument.
+*/
++ (JSValue *)valueWithNewBigIntFromString:(NSString *)string inContext:(JSContext *)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
These should probably have nullability attributes. Unfortunately, I think that means you'll have to add `_Null_unspecified` to every other pointer type in this file. So this should be:
`+ (JSValue * _Nullable)valueWithNewBigIntFromString:(NSString * _Nonnull)string inContext:(JSContext * _Nonnull)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));`

hyjorc1 replied:
I understand we need to add `_Null_unspecified` for other existing API. But still seems a bit weird since only the new APIs have specific nullable annotations.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -159,6 +159,50 @@ + (JSValue *)valueWithNewSymbolFromDescription:(NSString *)description inContext
+    JSValueRef bigInt = JSBigIntCreateWithString([context JSGlobalContextRef], OpaqueJSString::tryCreate(string).get(), &exception);
+    if (exception) {
+        [context notifyException:exception];
+        return [JSValue valueWithUndefinedInContext:context];
```
Maybe we should return `nullptr` here? Returning `undefined` is a bit confusing. It's probably worth changing the `valueWithNewPromiseInContext` return value too. I don't think that's used enough to be a bin compatibly risk.

hyjorc1 replied:
Would that cause any issues for the existing `valueWithNewPromiseInContext` applications since it's public?

kmiller68 replied:
It's possible but unlikely since we return nullptr in other places the most consistent way to check for exceptions is to check the context.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -159,6 +159,50 @@ + (JSValue *)valueWithNewSymbolFromDescription:(NSString *)description inContext
+    JSValueRef bigInt = JSBigIntCreateWithInt64([context JSGlobalContextRef], int64, &exception);
+    if (exception) {
+        [context notifyException:exception];
+        return [JSValue valueWithUndefinedInContext:context];
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -159,6 +159,50 @@ + (JSValue *)valueWithNewSymbolFromDescription:(NSString *)description inContext
+    JSValueRef bigInt = JSBigIntCreateWithUInt64([context JSGlobalContextRef], uint64, &exception);
+    if (exception) {
+        [context notifyException:exception];
+        return [JSValue valueWithUndefinedInContext:context];
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.mm:
```diff
@@ -159,6 +159,50 @@ + (JSValue *)valueWithNewSymbolFromDescription:(NSString *)description inContext
+    JSValueRef bigInt = JSBigIntCreateWithNumber([context JSGlobalContextRef], number, &exception);
+    if (exception) {
+        [context notifyException:exception];
+        return [JSValue valueWithUndefinedInContext:context];
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+
+/*!
+@method
+@abstract Create a new, unique, bigint object.
```
I think this should be "Create a new bigint value from a numeric string"

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+/*!
+@method
+@abstract Create a new, unique, bigint object.
+@param string The string representation of the bigint object being created.
```
Maybe say "The string representation of the bigint JavaScript value being created." since it's technically not an object in JS.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a unique JavaScript value with type bigint.
+*/
++ (JSValue * _Nullable)valueWithNewBigIntFromInt64:(int64_t)int64 inContext:(JSContext * _Nonnull)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Is int64_t in Obj-C or Obj-C++? I think API has to be Obj-C only.

kmiller68 replied:
At least https://developer.apple.com/documentation/foundation/nsnumber/1416870-longlongvalue?language=objc uses long long. Not sure if that means we should change the name of this function? We should also make sure the name uses int64 for Swift like it does for NSNumber if so.

hyjorc1 replied:
> Is int64_t in Obj-C or Obj-C++?

It seems that Obj-C has `uint64_t` and `int64_t`. 
https://developer.apple.com/documentation/kernel/uint64_t/
https://developer.apple.com/documentation/kernel/int64_t/

hyjorc1 replied:
For swift, we have `int64` and `uint64`.
https://developer.apple.com/documentation/swift/int64/
https://developer.apple.com/documentation/swift/uint64/

kmiller68 replied:
Ok, did you verify you can build with just the umbrella header included for Obj-c. I think we have a test for C but not Obj-C

hyjorc1 replied:
We have added test cases for `valueWithNewBigIntFromInt64`, `valueWithNewBigIntFromUInt64`, `valueWithNewBigIntFromString`, and `valueWithNewBigIntFromNumber` in `testapi.mm` which imports the umbrella header `JavaScriptCore.h`. So, I think it should be fine.

hyjorc1 replied:
OK, there is `testIncludes.m`.
```
// umbrella header
#import <JavaScriptCore/JavaScriptCore.h>
```

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a unique JavaScript value with type bigint.
+*/
++ (JSValue * _Nullable)valueWithNewBigIntFromUInt64:(uint64_t)uint64 inContext:(JSContext * _Nonnull)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Ditto for uint64_t.

hyjorc1 replied:
It seems that Obj-C has uint64_t and int64_t.
https://developer.apple.com/documentation/kernel/uint64_t/
https://developer.apple.com/documentation/kernel/int64_t/

hyjorc1 replied:
For swift, we have `int64` and `uint64`.
https://developer.apple.com/documentation/swift/int64/
https://developer.apple.com/documentation/swift/uint64/

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+
+/*!
+@method
+@abstract Create a new, unique, bigint object.
```
I think this should be "Create a new bigint value from an int64"

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+
+/*!
+@method
+@abstract Create a new, unique, bigint object.
```
I think this should be "Create a new bigint value from a uint64"

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,61 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+
+/*!
+@method
+@abstract Create a new, unique, bigint object.
```
I think this should be "Create a new bigint value from a double"

kmiller68 reviewed (commented):
Overall, LGTM

kmiller68 commented on Source/JavaScriptCore/runtime/JSCJSValue.h:
```diff
@@ -297,6 +297,7 @@ class JSValue {
     JSValue toNumeric(JSGlobalObject*) const;
     JSValue toBigIntOrInt32(JSGlobalObject*) const;
     JSBigInt* asHeapBigInt() const;
+    JSBigInt* asBigInt(JSGlobalObject*) const;
```
Dead code.

kmiller68 commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -606,6 +606,22 @@ ALWAYS_INLINE JSBigInt* JSValue::asHeapBigInt() const
+    ASSERT(isHeapBigInt());
+    return asHeapBigInt();
+}
+
```
Dead code.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -88,6 +92,21 @@ typedef enum {
+    kJSRelationConditionUndefined JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA)),
+    kJSRelationConditionEqual JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA)),
+    kJSRelationConditionGreaterThan JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA)),
+    kJSRelationConditionLessThan JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA))
```
I don't think you need the `JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA)),` since it's on the enum as a whole. You only need it if you added new cases after the initial declaration.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,133 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param right      The uint64_t as the right operand.
+    @result           A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if the value is not a BigInt.
+*/
+JS_EXPORT JSRelationCondition JSBigIntCompareUInt64(JSContextRef ctx, JSValueRef left, uint64_t right) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
I feel like if we're going to add relations we should make the API:
```
JSValueCompare(JSContextRef ctx, JSValueRef left, JSValueRef right, JSValueRef* exception);
JSValueCompareUInt64(JSContextRef ctx, JSValueRef left, uint64_t right, JSValueRef* exception);
JSValueCompareInt64(JSContextRef ctx, JSValueRef left, int64_t right, JSValueRef* exception);
JSValueCompareDouble(JSContextRef ctx, JSValueRef left, double right, JSValueRef* exception);
``` 
Since you can compare a number and BigInt in JS. Additionally, it's weird if we only support comparisons with BigInts and have no way to do it with other `JSValueRef`s.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,133 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @result           A double with the result of conversion, or 0 if an exception is thrown.
+    @discussion       This is equivalent to calling the `Number` constructor from JavaScript with a BigInt. If the value is not a BigInt, an exception is thrown.
+*/
+JS_EXPORT double JSBigIntToDouble(JSContextRef ctx, JSValueRef value, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Is there a reason to not just make BigInts work with `JSValueToNumber`?

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,133 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @result           A uint64_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       If the value is not a BigInt, an exception is thrown.
+*/
+JS_EXPORT uint64_t JSBigIntToUInt64(JSContextRef ctx, JSValueRef value, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
It seems like this could just be `JSValueToUInt64` and work with any `JSValueRef` now that I think about it.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,133 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @result           A int64_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       If the value is not a BigInt, an exception is thrown.
+*/
+JS_EXPORT int64_t JSBigIntToInt64(JSContextRef ctx, JSValueRef value, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,133 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param radix      An integer in the range 2 to 36 to be used as the base for representing the BigInt value.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A JSString with the result of conversion, or NULL if an exception is thrown. Ownership follows the Create Rule.
+    @discussion       This is equivalent to calling the `BigInt::toString` from JavaScript with a radix. If the value is not a BigInt or the radix is not in range 2 to 36, an exception is thrown.
```
Might also be worth mentioning that `JSBigIntToString` with a radix of `10` is the same as calling `JSValueToString`

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +376,70 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Truncate a JavaScript BigInt to a <code>uint64_t</code>.
+@result The uint64_t result of the truncation, 0 is returned if the JSValue is not a BigInt.
+*/
+- (uint64_t)toBigUInt64 JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
It seems like this could just be `toUInt64` and work with other `JSValue`s.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +376,70 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Truncate a JavaScript BigInt to a <code>int64_t</code>.
+@result The int64_t result of the truncation, 0 is returned if the JSValue is not a BigInt.
+*/
+- (int64_t)toBigInt64 JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +376,70 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@result The double result of the truncation, 0 is returned if the JSValue is not a BigInt.
+@discussion This is equivalent to calling the `Number` constructor from JavaScript with a BigInt.
+*/
+- (double)toBigIntDouble JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
I think we could just make this work with `toDouble`

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +376,70 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@other The uint64_t to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if this JSValue is not a BigInt.
+*/
+- (JSRelationCondition)compareBigIntUInt64:(uint64_t)other JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
Ditto with compare for `JSValueRef`

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,60 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Create a new BigInt value from a numeric string.
+@param string The string representation of the BigInt JavaScript value being created.
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a unique JavaScript value with type BigInt.
```
nit: I don't think these are "unique". The symbol API is unique because the same parameters should never give you the same JSValue back. We're not making that promise here.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
-@discussion The JSValue is converted to a number according to the rules specified 
- by the JavaScript language.
 @result The double result of the conversion.
+@discussion The result is computed by `ToNumeric` in JavaScript. If the JSValue is a BigInt, this is equivalent to `Number(value)` in JavaScript.
```
This is listing spec terminology. I think most users of the API aren't going to know what that means. I think this should be something like: "Convert the JSValue to a number according to the rules specified by the JavaScript language. Unless the JSValue is a BigInt then this is equivalent to `Number(value)` in JavaScript."

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @abstract Convert a JSValue to an <code>int32_t</code>.
-@discussion The JSValue is converted to an integer according to the rules specified 
- by the JavaScript language.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigInt64` and truncate it to a int32_t if the numeric is a BigInt, otherwise do `ToInt32`.
```
Maybe "The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to an `int32_t`"?

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @abstract Convert a JSValue to a <code>uint32_t</code>.
-@discussion The JSValue is converted to an integer according to the rules specified 
- by the JavaScript language.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigUint64` and truncate it to a uint32_t if the numeric is a BigInt, otherwise do `ToUint32`.
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/tests/testapi.cpp:
```diff
@@ -744,6 +759,515 @@ void TestAPI::testJSObjectSetOnGlobalObjectSubclassDefinition()
+
+    {
+        double number = 1.1;
+        const char* description = "checking JSValueMakeBigIntFromNumber with NaN";
```
Nit: 1.1 not NaN.

kmiller68 commented on Source/JavaScriptCore/API/tests/testapi.cpp:
```diff
@@ -744,6 +759,515 @@ void TestAPI::testJSObjectSetOnGlobalObjectSubclassDefinition()
+    WTF_TZONE_INIT(seed);
+    JSC::registerTZoneTypes();
+    WTF_TZONE_REGISTRATION_DONE();
+}
```
Why do we need this?

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@result The NSString result of the conversion, a <code>nil</code> is returned if the JSValue is not a BigInt or the radix is not in the range 2 to 36.
+@discussion This is equivalent to calling the `BigInt::toString` from JavaScript with a radix, and equivalent to `toString` if the radix is 10.
+*/
+- (nullable NSString *)toBigIntString:(size_t)radix JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
I wonder if this is too obscure to be worth adding. How often do people want then BigInt's stringified with radix != 10 that they can't just call a script with `BigInt.toString()`. Also, I think it would be `BigInt.toString` from JavaScript not `BigInt::toString`

hyjorc1 replied:
Yeah, let's remove them.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@method
+@abstract Compare a JSValue with a <code>uint64_t</code>.
+@other The uint64_t to compare with.
+@result A value of JSRelationCondition.
```
Might be worth a comment that "a kJSRelationConditionUndefined is returned if an exception is throw" for these.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,152 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @discussion       This is equivalent to calling the `BigInt::toString` from JavaScript with a radix, and equivalent to `JSValueToStringCopy` if the radix is 10. If the value is not a BigInt or the radix is not in range 2 to 36, an exception is thrown.
+*/
+JS_EXPORT JSStringRef JSBigIntToString(JSContextRef ctx, JSValueRef value, size_t radix, JSValueRef* exception) JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
+
```
Ditto on the necessity of this.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @abstract Convert a JSValue to a <code>uint32_t</code>.
-@discussion The JSValue is converted to an integer according to the rules specified 
- by the JavaScript language.
+@discussion The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to an `uint32_t`.
```
Typo: an `uint32_t` => a `uint32_t`

hyjorc1 replied:
Right, "a  yoo-int32_t". English is fun.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+/*!
+@method
+@abstract Convert a JSValue to a <code>int64_t</code>.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigInt64` if the numeric is a BigInt, otherwise do the truncation similar to `ToInt32` but for int64.
```
I think using spec language is going to be confusing and could possibly get outdated. I think we should update these comments to something like:
The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the value is a BigInt then the value is truncated to an `int64_t`

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+/*!
+@method
+@abstract Convert a JSValue to a <code>uint64_t</code>.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigUint64` if the numeric is a BigInt, otherwise do the truncation similar to `ToUint32` but for int64.
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,52 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
- to the rules specified by the JavaScript language.
+@discussion If the JSValue represents a boolean, a NSNumber value of YES or NO
+ will be returned. For all other types, the result is computed by `ToNumeric`
+ in JavaScript. If the JSValue is a BigInt, this is equivalent to `Number(value)` in JavaScript.
```
I think you can just say, "For all other types, the result is equivalent to `Number(value)` in JavaScript."

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,43 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Compare a JSValue with a <code>uint64_t</code>.
+@other The uint64_t to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the integer.
```
I would say "The JSValue is converted to an integer according to the rules specified by the JavaScript language then compared with `other`."

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,43 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Compare a JSValue with a <code>int64_t</code>.
+@other The int64_t to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the integer.
```
I would say "The JSValue is converted to an integer according to the rules specified by the JavaScript language then compared with `other`."

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,43 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Compare a JSValue with a double.
+@other The double to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
+@discussion Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the double.
```
I would say "The JSValue is converted to a double according to the rules specified by the JavaScript language then compared with `other`."

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -339,7 +389,43 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Compares two JSValues.
+@other The JSValue to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
+@discussion The result is computed by `IsLooselyEqual` and `IsLessThan` in JavaScript. If either the left or right JSValue has a NaN result of `ToPrimitive(value, NUMBER)`, a kJSRelationConditionUndefined is returned.
```
The result is computed by comparing the results of `this == other`, `this < other`, and `this > other`. If either `this` or `other` is (or would coerce to) `NaN` then the result is `kJSRelationConditionUndefined`

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A uint64_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigUint64` if the numeric is a BigInt, otherwise do the truncation similar to `ToUint32` but for int64.
```
The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the value is a BigInt then the value is truncated to a `uint64_t`

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A int64_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigInt64` if the numeric is a BigInt, otherwise do the truncation similar to `ToInt32` but for int64.
```
The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the value is a BigInt then the value is truncated to an `int64_t`

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A uint32_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigUint64` and truncate it to a uint32_t if the numeric is a BigInt, otherwise do `ToUint32`.
```
The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to a `uint32_t`.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A int32_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first. Then apply `ToBigInt64` and truncate it to a int32_t if the numeric is a BigInt, otherwise do `ToInt32`.
```
The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to an `int32_t`.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -332,6 +494,7 @@ JS_EXPORT bool JSValueToBoolean(JSContextRef ctx, JSValueRef value);
 @param value    The JSValue to convert.
 @param exception A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
 @result         The numeric result of conversion, or NaN if an exception is thrown.
+@discussion     The result is computed by `ToNumeric` in JavaScript. If the JSValue is a BigInt, this is equivalent to `Number(value)` in JavaScript.
```
The result is equivalent to `Number(value)` in JavaScript.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param right      The uint64_t as the right operand.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the integer.
```
I would say "`left` is converted to an integer according to the rules specified by the JavaScript language then compared with `right`."

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param right      The int64_t as the right operand.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the integer.
```
I would say "`left` is converted to an integer according to the rules specified by the JavaScript language then compared with `right`."

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param right      The double as the right operand.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is thrown.
+    @discussion       Convert the JSValue to a numeric by `ToNumeric` in JavaScript first and then compare with the double.
```
I would say "`left` is converted to a double according to the rules specified by the JavaScript language then compared with `right`."

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -292,6 +320,140 @@ JS_EXPORT JSValueRef JSValueMakeString(JSContextRef ctx, JSStringRef string);
+    @param right      The JSValue as the right operand.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is thrown.
+    @discussion       The result is computed by `IsLooselyEqual` and `IsLessThan` in JavaScript. If either the left or right JSValue has a NaN result of `ToPrimitive(value, NUMBER)`, a kJSRelationConditionUndefined is returned.
```
The result is computed by comparing the results of `left == right`, `left < right`, and `left > right` in JavaScript. If either `left` or `right` is (or would coerce to) `NaN` then the result is `kJSRelationConditionUndefined`.

kmiller68 reviewed (approved):
r=me with discussion edits. I would also put the declaration of `JSValueCompare` / `[compare:JSValue*]` first as I would expect those to be the most commonly used.

MenloDorian commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -169,23 +169,60 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@param context The JSContext to which the resulting JSValue belongs.
+@result The JSValue representing a JavaScript value with type BigInt.
+*/
++ (nullable JSValue *)valueWithNewBigIntFromDouble:(double)value inContext:(nonnull JSContext *)context JSC_API_AVAILABLE(macos(JSC_MAC_TBA), ios(JSC_IOS_TBA));
```
What happens if the input double value contains a fraction part?  Does this API truncate or round in some way?  Or does it error out and refuse to create a BigInt?  If it errors out, how is the error manifested?  I think this API needs work on clarifying these details.

MenloDorian replied:
Secondly, what happens if we encounter an OOME while trying to create a BigInt?  How is that represented?  I see that the pre-existing `valueWithNewSymbolFromDescription` didn't specify this either.  Is it specified elsewhere?

hyjorc1 replied:
> What happens if the input double value contains a fraction part? Does this API truncate or round in some way? Or does it error out and refuse to create a BigInt? If it errors out, how is the error manifested? I think this API needs work on clarifying these details.

`valueWithNewBigIntFromDouble` is implemented by using `JSBigIntCreateWithDouble` which has a discussion comment:
```
@discussion If the value is not an integer, an exception is thrown.
```

Yes, we should add that comment for `valueWithNewBigIntFromDouble` too. And it's defined by spec:
https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-bigint-constructor-number-value
https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-numbertobigint
So, no truncation is needed for converting from double to bigint as long as the double is an integer.

hyjorc1 replied:
> Secondly, what happens if we encounter an OOME while trying to create a BigInt? How is that represented? I see that the pre-existing valueWithNewSymbolFromDescription didn't specify this either. Is it specified elsewhere?

If any exceptions are thrown, the `context` will notify the exception. So, users should check whether the passed `context` got an exception or not after calling `valueWithNewBigIntFromString`. As well as other APIs.

MenloDorian commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,49 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
-@discussion The JSValue is converted to a number according to the rules specified 
- by the JavaScript language.
 @result The double result of the conversion.
+@discussion Convert the JSValue to a number according to the rules specified by the JavaScript language. Unless the JSValue is a BigInt then this is equivalent to <code>Number(value)</code> in JavaScript.
```
This discussion is ambiguous about what happens if the JSValue is a BigInt.  It doesn't say anything about it.  We should fix that.  According to https://github.com/tc39/proposal-bigint/blob/master/ADVANCED.md#no-implicit-conversions-or-mixed-operands, the VM would throw a TypeError.  How is that manifested in the API?  Or is that documented elsewhere?

hyjorc1 replied:
Previously, `toDouble` or `toNumber` doesn't support for BigInt. Since we're introducing BigInt API in this patch and we also want to support the conversion from BigInt to double[1], instead of creating a new API like `toBigIntDouble/Number` we decided to extend the existing `toDouble/Number` for BigInt to avoid unnecessary complexity for users.

[1] https://tc39.es/ecma262/#sec-number-constructor-number-value

MenloDorian commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,49 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @abstract Convert a JSValue to an <code>int32_t</code>.
-@discussion The JSValue is converted to an integer according to the rules specified 
- by the JavaScript language.
+@discussion The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to an <code>int32_t</code>.
```
Please add a comma `,` after `then`.

MenloDorian commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,49 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
 @abstract Convert a JSValue to a <code>uint32_t</code>.
-@discussion The JSValue is converted to an integer according to the rules specified 
- by the JavaScript language.
+@discussion The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the JSValue is a BigInt then the value is truncated to a <code>uint32_t</code>.
```
Ditto: comma `,` here and below.

MenloDorian commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -263,39 +300,49 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
-@discussion If the JSValue represents a boolean, a NSNumber value of YES or NO 
- will be returned. For all other types the value will be converted to a number according 
- to the rules specified by the JavaScript language.
+@discussion If the JSValue represents a boolean, a NSNumber value of YES or NO
+ will be returned. For all other types, the result is equivalent to <code>Number(value)</code> in JavaScript.
```
The JS spec says that BigInts cannot be coerced into a Number.  How is that manifested here in the API?

hyjorc1 replied:
> The JS spec says that BigInts cannot be coerced into a Number. 

Yes, this is true. It's still possible to convert BigInt to double. See https://tc39.es/ecma262/#sec-number-constructor-number-value.

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -402,17 +450,59 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@method
+@abstract Compares two JSValues.
+@other The JSValue to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
```
Typo: throw => thrown. Can you fix the other places too?

kmiller68 commented on Source/JavaScriptCore/API/JSValue.h:
```diff
@@ -402,17 +450,59 @@ NS_CLASS_AVAILABLE(10_9, 7_0)
+@abstract Compares two JSValues.
+@other The JSValue to compare with.
+@result A value of JSRelationCondition, a kJSRelationConditionUndefined is returned if an exception is throw.
+@discussion The result is computed by comparing the results of <code>this == other</code>, <code>this < other</code>, and <code>this > other</code>. If either <code>this</code> or <code>other</code> is (or would coerce to) <code>NaN</code>, then the result is kJSRelationConditionUndefined.
```
Hmm, maybe we should just say "The result is computed by comparing the results of JavaScript's <code>==</code>, <code>\<</code>, and <code>\></code> operators. If either <code>self</code> or <code>other</code> is (or would coerce to) <code>NaN</code> in JavaScript, then the result is kJSRelationConditionUndefined." Since `this` isn't a keyword in Obj-C and `self` isn't a keyword in JS.

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -332,9 +450,54 @@ JS_EXPORT bool JSValueToBoolean(JSContextRef ctx, JSValueRef value);
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           A int32_t with the result of conversion, or 0 if an exception is thrown.
+    @discussion       The JSValue is converted to an integer according to the rules specified by the JavaScript language. If the value is a BigInt, then the JSValue is truncated to a int32_t.
```
Typo: a int32_t => an int32_t

kmiller68 reviewed (commented):
Found a few more typos. Also, proposed a different `@discussion` for `[JSValue compare]`

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -32,6 +32,27 @@
+#define _Null_unspecified
+#define _Nullable
+#define _Nonnull
+#endif
```
We probably shouldn't be clobbering the global names for these. Let's move this to `JSBase.h` and change these to:

```
#if JSC_OBJC_API_ENABLED
#define JSC_CF_ENUM(enumName, ...)       \
    typedef CF_ENUM(uint32_t, enumName) { \
        __VA_ARGS__                       \
    }
#define JSC_ASSUME_NONNULL_BEGIN _Pragma("clang assume_nonnull begin")
#define JSC_ASSUME_NONNULL_END _Pragma("clang assume_nonnull end")
#define JSC_NULL_UNSPECIFIED _Null_unspecified
#define JSC_NULLABLE _Nullable
#define JSC_NONNULL _Nonnull
#else
#define JSC_CF_ENUM(enumName, ...) \
    typedef enum {                  \
        __VA_ARGS__                 \
    } enumName
#define JSC_ASSUME_NONNULL_BEGIN
#define JSC_ASSUME_NONNULL_END
#define JSC_NULL_UNSPECIFIED
#define JSC_NULLABLE
#define JSC_NONNULL
#endif
```

kmiller68 commented on Source/JavaScriptCore/API/JSValueRef.h:
```diff
@@ -332,8 +475,55 @@ JS_EXPORT bool JSValueToBoolean(JSContextRef ctx, JSValueRef value);
+    @param ctx        The execution context to use.
+    @param value      The JSValue to convert.
+    @param exception  A pointer to a JSValueRef in which to store an exception, if any. Pass NULL if you do not care to store an exception.
+    @result           An int32_t with the result of conversion, or 0 if an exception is thrown.
```
Seems like these are all missing the "Since 0 is valid value, `exception` must be checked after the call." comments. Can you update them all?

kmiller68 reviewed (changes requested):
Can you change the nullability macros to JSC specific ones? This way we don't clobber the values for non-apple platforms.

Also, I think you missed a few comment updates.

kmiller68 reviewed (approved):
r=me.

hyjorc1 committed:
[WIP] Introduce BigInt Public C API in JavaScriptCore
https://bugs.webkit.org/show_bug.cgi?id=250511
rdar://104194532

Reviewed by Keith Miller.

JavaScriptCore framework provides the ability to evaluate JavaScript programs
within applications developed using Swift and Object-C. While JavaScriptCore
APIs support all JavaScript primitive types except for BigInt type. This patch
introduces BigInt public C APIs for type verification, type access, value
creation, value conversion, and value comparison.

* Source/JavaScriptCore/API/APICast.h:
(toRef):
(toBigIntRef):
* Source/JavaScriptCore/API/JSBase.h:
* Source/JavaScriptCore/API/JSBigIntRef.cpp: Added.
(JSBigIntCreateWithNumber):
(JSBigIntCreateWithString):
(JSBigIntToString):
* Source/JavaScriptCore/API/JSBigIntRef.h: Added.
* Source/JavaScriptCore/API/JSValue.h:
* Source/JavaScriptCore/API/JSValue.mm:
(-[JSValue isBigInt]):
* Source/JavaScriptCore/API/JSValueRef.cpp:
(JSValueGetType):
(JSValueIsBigInt):
(JSValueMakeBigInt):
(JSValueToBigInt):
* Source/JavaScriptCore/API/JSValueRef.h:
* Source/JavaScriptCore/API/JavaScript.h:
* Source/JavaScriptCore/API/tests/testapi.cpp:
(TestAPI::checkIsBigIntType):
(TestAPI::testBigInt):
(testCAPIViaCpp):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::JSBigInt::stringToBigInt):
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::asBigInt const):

Canonical link: https://commits.webkit.org/278275@main
